File: scripts/scrape_scholarships/scrapers/collegescholarships.py
"""
Scraper for collegescholarships.org - 23,000+ scholarships.
Respect robots.txt and add delays between requests.
"""
import re
import time
from urllib.parse import urljoin

import requests
from bs4 import BeautifulSoup

BASE_URL = "https://www.collegescholarships.org"
HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.9",
}


def parse_amount(text: str) -> int:
    """Extract numeric amount from 'Award $25,000' or 'Award -'."""
    if not text or "-" in text:
        return 0
    match = re.search(r"\$[\d,]+", text)
    if match:
        return int(match.group().replace("$", "").replace(",", ""))
    return 0


def parse_deadline(text: str) -> str:
    """Convert 'February 28' or 'March 31' to YYYY-MM-DD. Uses current year."""
    if not text or text.lower() in ("rolling", "none", "varies"):
        return "2026-12-31"
    months = {
        "january": "01", "jan": "01", "february": "02", "feb": "02",
        "march": "03", "mar": "03", "april": "04", "apr": "04",
        "may": "05", "june": "06", "jun": "06", "july": "07", "jul": "07",
        "august": "08", "aug": "08", "september": "09", "sep": "09", "sept": "09",
        "october": "10", "oct": "10", "november": "11", "nov": "11",
        "december": "12", "dec": "12",
    }
    text = text.strip().lower()
    for name, mm in months.items():
        if name in text:
            match = re.search(r"(\d{1,2})(?:st|nd|rd|th)?", text)
            dd = match.group(1).zfill(2) if match else "01"
            return f"2026-{mm}-{dd}"
    return "2026-12-31"


def scrape_collegescholarships(max_pages: int = 5, delay_sec: float = 1.0) -> list[dict]:
    """
    Scrape scholarship listings from collegescholarships.org.
    Returns list of dicts matching ExternalScholarshipItem format.
    """
    results = []
    seen_ids = set()

    for page in range(max_pages):
        url = f"{BASE_URL}/financial-aid/" if page == 0 else f"{BASE_URL}/financial-aid/?start={page * 25}"
        try:
            resp = requests.get(url, headers=HEADERS, timeout=15)
            resp.raise_for_status()
        except requests.RequestException as e:
            print(f"  Error fetching page {page + 1}: {e}")
            break

        soup = BeautifulSoup(resp.text, "html.parser")

        # Find scholarship blocks - typically in divs or articles
        blocks = soup.select(".scholarship-item, .result-item, article.scholarship, div.scholarship")
        if not blocks:
            # Fallback: look for blocks containing "Award" and "Deadline"
            blocks = soup.find_all(["div", "article"], class_=re.compile(r"scholarship|result|item", re.I))
        if not blocks:
            # Try splitting by common patterns in the raw HTML
            for block in soup.find_all(["div", "article", "li"]):
                text = block.get_text()
                if "Award" in text and "Deadline" in text and "More info" in text:
                    blocks = soup.find_all(block.name, recursive=False)
                    break

        # Parse: look for links to /financial-aid/scholarship/ and extract from parent/siblings
        detail_links = soup.select('a[href*="/financial-aid/scholarship/"]')
        for link in detail_links:
            href = link.get("href", "")
            slug = href.split("/scholarship/")[-1].strip("/").split("?")[0]
            if not slug or slug in seen_ids:
                continue
            if slug.startswith("sort/") or slug == "sort":
                continue

            parent = link.find_parent(["div", "article", "li", "td"])
            if not parent:
                continue
            text = parent.get_text(separator=" ", strip=True)
            if "Award" not in text or "Deadline" not in text:
                continue

            seen_ids.add(slug)
            title_el = parent.find(["h4", "h3", "h2"])
            title = title_el.get_text(strip=True) if title_el else slug.replace("-", " ").title()

            amt_match = re.search(r"Award\s*:\s*\$?([\d,]+)|Award\s+\$([\d,]+)", text, re.I)
            amount = int((amt_match.group(1) or amt_match.group(2) or "0").replace(",", "")) if amt_match else 0

            dl_match = re.search(r"Deadline\s+([A-Za-z][^.\n]*?)(?:\s+\[|$)", text)
            deadline = parse_deadline(dl_match.group(1).strip()) if dl_match else "2026-12-31"

            desc = ""
            for p in parent.find_all("p"):
                t = p.get_text(strip=True)
                if len(t) > 80 and ("Applicant" in t or "Selection" in t):
                    desc = t[:500]
                    break
            full_url = urljoin(BASE_URL, href)
            if not desc:
                desc = f"{title}. Apply at {full_url}"

            results.append({
                "id": f"collegescholarships-{slug}",
                "title": title,
                "sponsor": "CollegeScholarships.org",
                "amount": amount,
                "deadline": deadline,
                "description": desc,
            })

        # Fallback: regex over raw HTML if BeautifulSoup found nothing
        if len(results) == 0 and page == 0:
            for m in re.finditer(
                r'Award\s+\$?([\d,]*)\s+Deadline\s+([^<\n]+).*?href="([^"]*financial-aid/scholarship/([^"/?]+))"[^>]*>.*?<h[234][^>]*>([^<]+)',
                resp.text,
                re.DOTALL | re.I,
            ):
                slug = m.group(4)
                if slug in seen_ids:
                    continue
                seen_ids.add(slug)
                amount = int(m.group(1).replace(",", "")) if m.group(1) else 0
                deadline = parse_deadline(m.group(2).strip())
                url_path = m.group(3)
                title = m.group(5).strip()
                full_url = urljoin(BASE_URL, url_path) if url_path.startswith("/") else url_path
                results.append({
                    "id": f"collegescholarships-{slug}",
                    "title": title,
                    "sponsor": "CollegeScholarships.org",
                    "amount": amount,
                    "deadline": deadline,
                    "description": f"{title}. Apply at {full_url}",
                })

        print(f"  Page {page + 1}: {len(results)} total")
        time.sleep(delay_sec)

    return results

File: lib/scrapers/collegescholarships.ts
/**
 * Scraper for collegescholarships.org - 23,000+ scholarships.
 * Scrapes main financial-aid listing (paginated with ?page=N) and our-scholarships section.
 */
import * as cheerio from "cheerio";
import type { ScrapedScholarship } from "./types";
import { parseAmountFromText } from "./parseAmount";
import { SCRAPER_HEADERS, parseDeadline, delay } from "./shared";

const BASE_URL = "https://www.collegescholarships.org";

function extractFromListing($: cheerio.CheerioAPI, results: ScrapedScholarship[], seen: Set<string>, baseUrl: string): void {
  $('a[href*="/financial-aid/scholarship/"]').each((_, el) => {
    const href = $(el).attr("href") ?? "";
    const slug = href.split("/scholarship/")[1]?.split("?")[0]?.replace(/\/$/, "") ?? "";
    if (!slug || seen.has(slug) || slug.startsWith("sort/")) return;

    const parent = $(el).closest("div, article, li, td");
    if (!parent.length) return;
    const text = parent.text();
    if (!text.includes("Award") && !text.includes("Deadline") && !text.match(/\$[\d,]+/)) return;

    seen.add(slug);
    const titleEl = parent.find("h4, h3, h2").first();
    const title = titleEl.text().trim() || slug.replace(/-/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());

    let amount = parseAmountFromText(text);
    if (amount === 0) {
      const amtMatch = text.match(/Award\s*:\s*\$?([\d,]+)|Award\s+\$([\d,]+)/i);
      amount = amtMatch ? parseInt((amtMatch[1] || amtMatch[2] || "0").replace(/,/g, ""), 10) : 0;
    }

    const dlMatch = text.match(/Deadline\s+([A-Za-z][^.\n]*?)(?:\s+\[|$)/);
    const deadline = dlMatch ? parseDeadline(dlMatch[1].trim()) : "2026-12-31";

    let desc = "";
    const parts: string[] = [];
    parent.find("p").each((_, p) => {
      const t = $(p).text().trim();
      if (t.length > 60) parts.push(t);
    });
    parent.find("li").each((_, li) => {
      const t = $(li).text().trim();
      if (t.length > 5 && t.length < 150) parts.push(`• ${t}`);
    });
    if (parts.length > 0) desc = parts.join("\n\n").slice(0, 800);
    const fullUrl = href.startsWith("http") ? href : `${baseUrl}${href.startsWith("/") ? "" : "/"}${href}`;
    if (!desc) desc = `${title}. Apply at ${fullUrl}`;
    else if (!desc.includes("Apply at")) desc += `\n\nApply at ${fullUrl}`;

    results.push({
      id: `collegescholarships-${slug}`,
      title,
      sponsor: "CollegeScholarships.org",
      amount,
      deadline,
      description: desc,
    });
  });
}

export async function scrapeCollegeScholarships(maxPages = 10): Promise<ScrapedScholarship[]> {
  const results: ScrapedScholarship[] = [];
  const seen = new Set<string>();

  // Main financial-aid listing: ?page=N (1-based)
  for (let page = 1; page <= maxPages; page++) {
    try {
      const url = page === 1
        ? `${BASE_URL}/financial-aid/`
        : `${BASE_URL}/financial-aid/?page=${page}`;
      const res = await fetch(url, { headers: SCRAPER_HEADERS });
      if (!res.ok) break;
      const html = await res.text();
      const $ = cheerio.load(html);
      extractFromListing($, results, seen, BASE_URL);
      await delay(800);
    } catch (e) {
      console.error("[scraper] collegescholarships page failed:", page, e);
    }
  }

  // Our Scholarships section (different listing)
  try {
    const res = await fetch(`${BASE_URL}/our-scholarships/`, { headers: SCRAPER_HEADERS });
    if (res.ok) {
      const html = await res.text();
      const $ = cheerio.load(html);
      $('a[href*="/financial-aid/scholarship/"], a[href*="/scholarship/"]').each((_, el) => {
        const href = $(el).attr("href") ?? "";
        const slug = (href.match(/\/scholarship\/([^/?#]+)/) ?? [])[1]?.replace(/\/$/, "") ?? "";
        if (!slug || seen.has(slug) || slug.startsWith("sort/")) return;
        const parent = $(el).closest("div, article, li, p");
        const block = parent.length ? parent : $(el).parent();
        const text = block.text();
        if (!text.trim() || text.length < 10) return;
        seen.add(slug);
        const titleEl = block.find("h1, h2, h3, h4").first();
        const title = titleEl.text().trim() || slug.replace(/-/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
        const amount = parseAmountFromText(text);
        const deadline = parseDeadline(text);
        const fullUrl = href.startsWith("http") ? href : `${BASE_URL}${href.startsWith("/") ? "" : "/"}${href}`;
        results.push({
          id: `collegescholarships-${slug}`,
          title,
          sponsor: "CollegeScholarships.org",
          amount: amount || 1000,
          deadline,
          description: text.slice(0, 600).trim() || `${title}. Apply at ${fullUrl}`,
        });
      });
      await delay(800);
    }
  } catch (e) {
    console.error("[scraper] collegescholarships our-scholarships failed:", e);
  }

  return results;
}
